File: solution/260. Single Number III/260.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        long ab = 0;
        for (int x : nums)
            ab ^= x;
        
        int lastBit = ab & (-ab);
        
        vector<int> res = {0, 0};
        for (int x : nums) {
            if (x & lastBit)
                res[0] ^= x;
            else
                res[1] ^= x;
        }
        return res;
    }
};

File: solution/260. Single Number III/260.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        ab = reduce(lambda a, b: a ^ b, nums)
        
        last_bit = ab & (-ab)
        
        res = [0, 0]
        for x in nums:
            if x & last_bit:
                res[0] ^= x
            else:
                res[1] ^= x
        return res
    
